### DISCUSSION TITLE
pylon capture node issue

### DISCUSSION THREAD

**Discussion by emilyjanedennis** [bonsai-rx/bonsai#916] [Q&A]
hi folks,

we are using a Basler camera and the pylon capture node is consistently causing issues. When we only use the capture node, everything works fine, but adding processes later on (starting other completely independent processes simultaneously like a rtsp stream or triggering an arduino at regular intervals, or processing the data with deeplabcut) SOMETIMES but not always causes a weird behavior...

If we visualize the output from the node, it looks like the top ~100 rows of data are "live" while the other 1948  (it's a 2048x2048 image, 50fps) are either frozen or repeating every second or so. The camera itself is not having issues (if I use the Basler Pylon software, we never get this problem). The camera has a dedicated USB port on a dedicated PCIE, so it's not a USB conflict. If I downsize the images coming out of it to say 200x200px, it seems to work just fine most of the time. This feels like a memory allocation issue or thread competition problem, but we're only using ~8% of our CPU memory and with deeplabcut on, we use about 20% of our GPU RAM.

In the mean time, I can re-train DLC on downsampled images, but I'd like to understand why this problem occurs and what I could do to fix it. I'm also calling Basler tomorrow to see if they can help, will post if I get something useful out of them but in the past their customer service has been pretty unhelpful, so also posting here. 

Thanks!

**Comment by emilyjanedennis** [comment]
we've now downsampled to 256x256 and are having the same issue

**Comment by emilyjanedennis** [comment]
so we were able to solve this using just the basler software to separately set the camera parameters like exposure time, fps, etc. and then use the CameraCapture node instead of the Basler node. This does hint that the issue is not with the Basler camera or on board firmware but a problem with the node itself.

**Comment by lillysanchez** [comment]
I concur that there is something wrong with the PylonCapture node. I had very similar problems @emilyjanedennis. I checked the PylonViewer software, but that produced a clear picture without any lagging. The view from the PylonCapture module within Bonsai was lagging by a few seconds and had lines in the picture. Most of the coordinates output from a tracking algorithm (I am tracking an LED) was NaNs. 

However, using the CameraCapture didn't help me. The video (uncropped) was only half of what the camera was capturing in PylonViewer. I can't actually have the camera connected to Bonsai and the PylonView program at the same time.  

Do you have any suggestions for how to work around the issue?

**Comment by IamJuanHilario** [comment]
Hi @emilyjanedennis How is it that you were able to established communication between the Basler camera and Bonsai? When I try to add it using a camera capture node I cannot connect it because there is no index that the camera gets connected to. I will sincerely appreciate your input. Thank you in advance.

**Reply by chiyu1203** [reply]
Did you install pylon library on bonsai, pylon viewer 7.1.0 and use the node called pyloncapture? After that you need to create a feature file on pylon viewer first and load that information in the pyloncapture node.

**Reply by emilyjanedennis** [reply]
actually we use the cameracapture node as-is. On our system, pylon needs to be closed. For us this mapped to "1" and our onboard webcam was "0". Making sure your CameraCapture parameters match your pylon settings is also useful to get visualizations/saving to work well. Specifically, make sure the frame size is the same as your image dimensions. I never was able to get pylon capture to work.

**Reply by Sarruedi** [reply]
Has anyone tried using Pylon viewer 7.4? I'm trying to set up a Basler camera from scratch and also run into issue with Pylon capture. The error is: External component has thrown an exception. I can connect to the camera using Pylon Viewer and it seems to work fine, so I'm not sure what the issue is on the Bonsai end. I saved the parameter file from Pylon Viewer and load these into Bonsai and specify the serial number as well, but so far no success.

PS: I can actually connect to the camera after installing Pylon viewer 7.1 (no success with 5.1 and 7.4) but I have to use Pylon capture, Camera Capture doesn't seem to work for me.

**Reply by bruno-f-cruz** [reply]
I would install the version that the current package targets (7.1) and try again.

**Reply by Sarruedi** [reply]
Thanks Bruno, you are right, 7.1 does the trick!

**Reply by bruno-f-cruz** [reply]
CameraCapture uses a generic opencv backend. I suspect basler does not ship generic webcam drivers with their software. So I would probably stick with the camera specific node as it offers more functionality anyway
